File: spark-doris-connector/spark-doris-connector-base/src/main/scala/org/apache/doris/spark/client/read/AbstractThriftReader.scala
package org.apache.doris.spark.client.read

import org.apache.doris.sdk.thrift.{TScanCloseParams, TScanNextBatchParams, TScanOpenParams, TScanOpenResult}
import org.apache.doris.spark.client.{DorisBackend, DorisFrontend, DorisReaderPartition, DorisSchema}
import org.apache.doris.spark.config.{DorisConfig, DorisOptions}
import org.slf4j.{Logger, LoggerFactory}

import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.locks.{Condition, Lock, ReentrantLock}
import java.util.concurrent.{ArrayBlockingQueue, BlockingQueue, TimeUnit}
import scala.collection.JavaConverters._
import scala.util.control.Breaks

protected[spark] abstract class AbstractThriftReader(partition: DorisReaderPartition, config: DorisConfig)
  extends DorisReader(partition, config) {

  protected val log: Logger = LoggerFactory.getLogger(this.getClass.getName.stripSuffix("$"))

  protected lazy val frontend: DorisFrontend = DorisFrontend(config)
  private lazy val backend = new DorisBackend(partition.backend, config)

  private[this] var offset = 0

  private[this] val eos: AtomicBoolean = new AtomicBoolean(false)
  private[this] lazy val async: Boolean = config.getValue(DorisOptions.DORIS_DESERIALIZE_ARROW_ASYNC)

  private[this] val beLock = if (async) new ReentrantLock() else new NoOpLock

  private val scanOpenParams = buildScanOpenParams()
  protected val scanOpenResult: TScanOpenResult = lockClient(_.openScanner(scanOpenParams))
  protected val contextId: String = scanOpenResult.getContextId
  protected val dorisSchema: DorisSchema

  private[this] val rowBatchQueue: BlockingQueue[RowBatch] = {
    if (async) {
      val blockingQueueSize = config.getValue(DorisOptions.DORIS_DESERIALIZE_QUEUE_SIZE)
      new ArrayBlockingQueue(blockingQueueSize)
    } else null
  }

  private[this] val asyncThread: Thread = new Thread {
    override def run(): Unit = {
      val nextBatchParams = new TScanNextBatchParams
      nextBatchParams.setContextId(contextId)
      while (!eos.get) {
        nextBatchParams.setOffset(offset)
        val nextResult = lockClient(_.getNext(nextBatchParams))
        eos.set(nextResult.isEos)
        if (!eos.get) {
          rowBatch = new RowBatch(nextResult, dorisSchema)
          offset += rowBatch.getReadRowCount
          rowBatch.close()
          rowBatchQueue.put(rowBatch)
        }
      }
    }
  }

  private val asyncThreadStarted: Boolean = {
    var started = false
    if (async) {
      asyncThread.start()
      started = true
    }
    started
  }

  override def hasNext: Boolean = {
    var hasNext = false
    if (async && asyncThreadStarted) {
      // support deserialize Arrow to RowBatch asynchronously
      if (rowBatch == null || !rowBatch.hasNext) {
        val loop = new Breaks
        loop.breakable {
          while (!eos.get || !rowBatchQueue.isEmpty) {
            if (!rowBatchQueue.isEmpty) {
              rowBatch = rowBatchQueue.take
              hasNext = true
              loop.break
            } else {
              // wait for rowBatch put in queue or eos change
              Thread.sleep(5)
            }
          }
        }
      } else {
        hasNext = true
      }
    } else {
      // Arrow data was acquired synchronously during the iterative process
      if (!eos.get && (rowBatch == null || !rowBatch.hasNext)) {
        if (rowBatch != null) {
          offset += rowBatch.getReadRowCount
          rowBatch.close()
        }
        val nextBatchParams = new TScanNextBatchParams
        nextBatchParams.setContextId(contextId)
        nextBatchParams.setOffset(offset)
        val nextResult = lockClient(_.getNext(nextBatchParams))
        eos.set(nextResult.isEos)
        if (!eos.get) {
          rowBatch = new RowBatch(nextResult, dorisSchema)
        }
      }
      hasNext = !eos.get
    }
    hasNext
  }

  override def next(): AnyRef = {
    if (!hasNext) {
      throw new Exception()
    }
    rowBatch.next.toArray
  }

  override def close(): Unit = {
    val closeParams = new TScanCloseParams
    closeParams.setContextId(contextId)
    lockClient(_.closeScanner(closeParams))
  }

  private def buildScanOpenParams(): TScanOpenParams = {
    val params = new TScanOpenParams
    params.cluster = DorisOptions.DORIS_DEFAULT_CLUSTER
    params.database = partition.database
    params.table = partition.table
    params.tablet_ids = partition.tablets.map(java.lang.Long.valueOf).toList.asJava
    params.opaqued_query_plan = partition.opaquedQueryPlan

    // max row number of one read batch
    val batchSize = config.getValue(DorisOptions.DORIS_BATCH_SIZE)
    val queryDorisTimeout = config.getValue(DorisOptions.DORIS_REQUEST_QUERY_TIMEOUT_S)
    val execMemLimit = config.getValue(DorisOptions.DORIS_EXEC_MEM_LIMIT)

    params.setBatchSize(batchSize)
    params.setQueryTimeout(queryDorisTimeout)
    params.setMemLimit(execMemLimit)
    params.setUser(config.getValue(DorisOptions.DORIS_USER))
    params.setPasswd(config.getValue(DorisOptions.DORIS_PASSWORD))

    log.debug(s"Open scan params is, " +
      s"cluster: ${params.getCluster}, " +
      s"database: ${params.getDatabase}, " +
      s"table: ${params.getTable}, " +
      s"tabletId: ${params.getTabletIds}, " +
      s"batch size: $batchSize, " +
      s"query timeout: $queryDorisTimeout, " +
      s"execution memory limit: $execMemLimit, " +
      s"user: ${params.getUser}, " +
      s"query plan: ${params.getOpaquedQueryPlan}")
    params
  }

  private def lockClient[T](action: DorisBackend => T): T = {
    beLock.lock()
    try {
      action(backend)
    } finally {
      beLock.unlock()
    }
  }

  private class NoOpLock extends Lock {

    override def lock(): Unit = {}

    override def lockInterruptibly(): Unit = {}

    override def tryLock(): Boolean = true

    override def tryLock(time: Long, unit: TimeUnit): Boolean = true

    override def unlock(): Unit = {}

    override def newCondition(): Condition = {
      throw new UnsupportedOperationException("NoOpLock can't provide a condition")
    }

  }

}

File: spark-doris-connector/spark-doris-connector-base/src/main/java/org/apache/doris/spark/client/read/AbstractThriftReader.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.spark.client.read;

import org.apache.doris.sdk.thrift.TPrimitiveType;
import org.apache.doris.sdk.thrift.TScanBatchResult;
import org.apache.doris.sdk.thrift.TScanCloseParams;
import org.apache.doris.sdk.thrift.TScanNextBatchParams;
import org.apache.doris.sdk.thrift.TScanOpenParams;
import org.apache.doris.sdk.thrift.TScanOpenResult;
import org.apache.doris.spark.client.DorisBackendThriftClient;
import org.apache.doris.spark.client.DorisFrontendClient;
import org.apache.doris.spark.client.entity.DorisReaderPartition;
import org.apache.doris.spark.config.DorisOptions;
import org.apache.doris.spark.exception.ConnectedFailedException;
import org.apache.doris.spark.exception.DorisException;
import org.apache.doris.spark.exception.OptionRequiredException;
import org.apache.doris.spark.rest.models.Field;
import org.apache.doris.spark.rest.models.Schema;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Function;
import java.util.stream.Collectors;

public abstract class AbstractThriftReader extends DorisReader {

    protected final Logger logger = LoggerFactory.getLogger(this.getClass().getName().replace("$", ""));

    protected final DorisFrontendClient frontend;
    private final DorisBackendThriftClient backend;

    private int offset = 0;

    private final AtomicBoolean endOfStream = new AtomicBoolean(false);
    private final boolean isAsync;

    private final Lock backendLock;
    private final TScanOpenParams scanOpenParams;
    protected final TScanOpenResult scanOpenResult;
    protected final String contextId;
    protected Schema dorisSchema;

    private final BlockingQueue<RowBatch> rowBatchQueue;

    private final Thread asyncThread;

    private int readCount = 0;

    private final Boolean datetimeJava8ApiEnabled;

    protected AbstractThriftReader(DorisReaderPartition partition) throws Exception {
        super(partition);
        this.frontend = new DorisFrontendClient(config);
        this.backend = new DorisBackendThriftClient(partition.getBackend(), config);
        this.isAsync = config.getValue(DorisOptions.DORIS_DESERIALIZE_ARROW_ASYNC);
        this.backendLock = isAsync ? new ReentrantLock() : new NoOpLock();
        this.scanOpenParams = buildScanOpenParams();
        this.scanOpenResult = lockClient(backend -> {
            try {
                return backend.openScanner(scanOpenParams);
            } catch (ConnectedFailedException e) {
                throw new RuntimeException("backend open scanner failed", e);
            }
        });
        this.contextId = scanOpenResult.getContextId();
        Schema schema = getDorisSchema();
        this.dorisSchema = processDorisSchema(partition, schema);
        if (logger.isDebugEnabled()) {
            logger.debug("origin thrift read Schema: " + schema + ", processed schema: " + dorisSchema);
        }
        if (isAsync) {
            int blockingQueueSize = config.getValue(DorisOptions.DORIS_DESERIALIZE_QUEUE_SIZE);
            this.rowBatchQueue = new ArrayBlockingQueue<>(blockingQueueSize);
            this.asyncThread = new Thread(() -> {
                try {
                    runAsync();
                } catch (DorisException | InterruptedException e) {
                    throw new RuntimeException("backend read async failed", e);
                }
            });
            this.asyncThread.start();
        } else {
            this.rowBatchQueue = null;
            this.asyncThread = null;
        }
        this.datetimeJava8ApiEnabled = partition.getDateTimeJava8APIEnabled();
    }

    private void runAsync() throws DorisException, InterruptedException {
        TScanNextBatchParams nextBatchParams = new TScanNextBatchParams();
        nextBatchParams.setContextId(contextId);
        while (!endOfStream.get()) {
            nextBatchParams.setOffset(offset);
            TScanBatchResult nextResult = lockClient(backend -> {
                try {
                    return backend.getNext(nextBatchParams);
                } catch (DorisException e) {
                    throw new RuntimeException("backend async get next failed", e);
                }
            });
            endOfStream.set(nextResult.isEos());
            if (!endOfStream.get()) {
                rowBatch = new RowBatch(nextResult, dorisSchema, datetimeJava8ApiEnabled);
                offset += rowBatch.getReadRowCount();
                rowBatch.close();
                rowBatchQueue.put(rowBatch);
            }
        }
    }

    @Override
    public boolean hasNext() throws DorisException {
        if (partition.getLimit() > 0 && readCount >= partition.getLimit()) {
            return false;
        }
        boolean hasNext = false;
        if (isAsync && asyncThread != null && asyncThread.isAlive()) {
            if (rowBatch == null || !rowBatch.hasNext()) {
                while (!endOfStream.get() || !rowBatchQueue.isEmpty()) {
                    if (!rowBatchQueue.isEmpty()) {
                        try {
                            rowBatch = rowBatchQueue.take();
                            hasNext = true;
                            break;
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                        }
                    } else {
                        try {
                            Thread.sleep(5);
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                        }
                    }
                }
            } else {
                hasNext = true;
            }
        } else {
            if (!endOfStream.get() && (rowBatch == null || !rowBatch.hasNext())) {
                if (rowBatch != null) {
                    offset += rowBatch.getReadRowCount();
                    rowBatch.close();
                }
                TScanNextBatchParams nextBatchParams = new TScanNextBatchParams();
                nextBatchParams.setContextId(contextId);
                nextBatchParams.setOffset(offset);
                TScanBatchResult nextResult = lockClient(backend -> {
                    try {
                        return backend.getNext(nextBatchParams);
                    } catch (DorisException e) {
                        throw new RuntimeException("backend get next failed", e);
                    }
                });
                endOfStream.set(nextResult.isEos());
                if (!endOfStream.get()) {
                    rowBatch = new RowBatch(nextResult, dorisSchema, datetimeJava8ApiEnabled);
                }
            }
            hasNext = !endOfStream.get();
        }
        return hasNext;
    }

    @Override
    public Object next() throws DorisException {
        if (!hasNext()) {
            throw new RuntimeException("No more elements");
        }
        if (partition.getLimit() > 0) {
            readCount++;
        }
        return rowBatch.next().toArray();
    }

    @Override
    public void close() {
        TScanCloseParams closeParams = new TScanCloseParams();
        closeParams.setContextId(contextId);
        lockClient(backend -> {
            backend.closeScanner(closeParams);
            return null;
        });
    }

    private TScanOpenParams buildScanOpenParams() throws OptionRequiredException {
        TScanOpenParams params = new TScanOpenParams();
        params.setCluster(DorisOptions.DORIS_DEFAULT_CLUSTER);
        params.setDatabase(partition.getDatabase());
        params.setTable(partition.getTable());
        List<Long> tabletIds = new ArrayList<>(Arrays.asList(partition.getTablets()));
        params.setTabletIds(tabletIds);
        params.setOpaquedQueryPlan(partition.getOpaquedQueryPlan());

        int batchSize = config.getValue(DorisOptions.DORIS_BATCH_SIZE);
        if (batchSize > DorisOptions.DORIS_BATCH_SIZE_MAX) {
            batchSize = DorisOptions.DORIS_BATCH_SIZE_MAX;
        }
        int queryDorisTimeout = config.getValue(DorisOptions.DORIS_REQUEST_QUERY_TIMEOUT_S);
        long execMemLimit = config.getValue(DorisOptions.DORIS_EXEC_MEM_LIMIT);

        params.setBatchSize(batchSize);
        params.setQueryTimeout(queryDorisTimeout);
        params.setMemLimit(execMemLimit);
        params.setUser(config.getValue(DorisOptions.DORIS_USER));
        params.setPasswd(config.getValue(DorisOptions.DORIS_PASSWORD));

        logger.debug("Open scan params: cluster: {}, database: {}, table: {}, tabletId: {}, batch size: {}, query timeout: {}, execution memory limit: {}, user: {}, query plan: {}",
                params.getCluster(), params.getDatabase(), params.getTable(), params.getTabletIds(), batchSize, queryDorisTimeout, execMemLimit, params.getUser(), params.getOpaquedQueryPlan());
        return params;
    }

    private <T> T lockClient(Function<DorisBackendThriftClient, T> action) {
        backendLock.lock();
        try {
            return action.apply(backend);
        } finally {
            backendLock.unlock();
        }
    }

    protected abstract Schema getDorisSchema() throws Exception;

    protected Schema processDorisSchema(DorisReaderPartition partition, final Schema originSchema) throws Exception {
        Schema processedSchema = new Schema();
        Schema tableSchema = frontend.getTableSchema(partition.getDatabase(), partition.getTable());
        Map<String, Field> fieldTypeMap = tableSchema.getProperties().stream()
                .collect(Collectors.toMap(Field::getName, Function.identity()));
        Map<String, Field> scanTypeMap = originSchema.getProperties().stream()
                .collect(Collectors.toMap(Field::getName, Function.identity()));
        String[] readColumns = partition.getReadColumns();
        List<Field> newFieldList = new ArrayList<>();
        for (String readColumn : readColumns) {
            if (readColumn.contains(" AS ")) {
                int asIdx = readColumn.indexOf(" AS ");
                String realColumn = readColumn.substring(asIdx + 4).trim().replaceAll("`", "");
                if (fieldTypeMap.containsKey(realColumn) && scanTypeMap.containsKey(realColumn)
                        && ("BITMAP".equalsIgnoreCase(fieldTypeMap.get(realColumn).getType())
                        || "HLL".equalsIgnoreCase(fieldTypeMap.get(realColumn).getType()))) {
                    newFieldList.add(new Field(realColumn, TPrimitiveType.VARCHAR.name(), null, 0, 0, null));
                }
            } else {
                newFieldList.add(scanTypeMap.get(readColumn.trim().replaceAll("`", "")));
            }
        }
        processedSchema.setProperties(newFieldList);
        return processedSchema;
    }

    private static class NoOpLock implements Lock {
        @Override
        public void lock() {}

        @Override
        public void lockInterruptibly() {}

        @Override
        public boolean tryLock() {
            return true;
        }

        @Override
        public boolean tryLock(long time, TimeUnit unit) {
            return true;
        }

        @Override
        public void unlock() {}

        @Override
        public Condition newCondition() {
            throw new UnsupportedOperationException("NoOpLock can't provide a condition");
        }
    }
}